### ISSUE TITLE
`no-unused-vars` should not flag constants declared with `const`

### ISSUE THREAD

**Issue by bluesmoon** [eslint/eslint#6131]
**What version of ESLint are you using?**
2.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```json
{
	"env": {
		"browser":true,
		"es6":true
	},
	"plugins": [
		"html"
	],
	"rules": {
		"no-mixed-spaces-and-tabs":[2, true],
		"no-redeclare":2,
		"no-use-before-define":[2, "nofunc"],
		"no-unused-vars":1,
		"no-undef":2,
		"eqeqeq":2,
		"consistent-return":1,
		"radix":2,
		"console":0,
		"camelcase":0,
		"strict":0,
		"quotes":[2, "double", "avoid-escape"],
		"new-cap":0,
		"space-infix-ops":0,
		"no-console":0,
		"no-delete-var":0,
		"no-underscore-dangle":0,
		"no-multi-spaces":0,
		"dot-notation":[2, {"allowKeywords": false}],
		"space-unary-ops":1,
		"key-spacing": [1, {"beforeColon": false, "afterColon": true, "mode": "minimum"}],
		"no-empty":2,
		"brace-style": [1, "stroustrup", { "allowSingleLine": true }],
		"keyword-spacing": 2,
		"no-space-before-semi": 0,
		"space-before-function-paren": [2, "never"],
		"no-trailing-spaces": [2, { "skipBlankLines": false }],
		"prefer-const": ["error"]
	}
}
```

**What did you do? Please include the actual source code causing the issue.**
```javascript
const EDGE_OFFSET = 15;
```

**What did you expect to happen?**
No warnings since `EDGE_OFFSET` is a constant, not a variable.

**What actually happened? Please include the actual, raw output from ESLint.**
```
minimum.js
  1:7  warning  'EDGE_OFFSET' is defined but never used  no-unused-vars

✖ 1 problem (0 errors, 1 warning)
```

**Comment by ilyavolodin** [comment]
Why do you think constants shouldn't be flagged? If you never use the value of the const, you can safely remove it from your code.

**Comment by bluesmoon** [comment]
a const isn't a variable.  It should not be flagged with `no-unused-vars` which is meant for variables.

**Comment by ilyavolodin** [comment]
So you are arguing semantics of the name of the rule? Const is a read-only variable (not really completely true in JS, but that's what it is in other languages).

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by ghengeveld** [comment]
There's two ways you can avoid this error:

1. `export` your const. That's what you probably intended to do anyway, assuming you use modules.
2. Configure ESLint to ignore these errors. In this case you could use a regex to ignore CONSTANT_CASE variable names:

```json
"no-unused-vars": ["error", { "varsIgnorePattern": "[A-Z_]+" }]
```

However you should note that this will also ignore such names in other places.

**Comment by bluesmoon** [comment]
@ghengeveld thanks, this is good since by convention only (and all) `const`s use all caps in our code

**Comment by ghengeveld** [comment]
Of course I'm not familiar with your codebase and conventions, but nowadays it's common to use `const` everywhere, not just for actual (global) constants. If I need a `let` (or `var`) I will rewrite my code to not need a `let`. In general I write all my code in an immutable fashion, so `const` is the only way to go. This means not just my contants (for example [Redux action types](http://redux.js.org/docs/basics/Actions.html)) will use `const`, but also any other variables.

What I'm trying to say is, don't hold on too tight to your ALL_CAPS convention. It generally doesn't help readability if your code is sprinkled with all caps.
